test/specs/EventsSpec.js

```diff
@@ -21,325 +21,355 @@ describe("Event Tests", function() {
     });
 
     describe("Mouse Events", function() {
+      var spy;
 
-      it("'slideStart' event is triggered properly and can be binded to", function() {
-        testSlider.on('slideStart', function() {
-          flag = true;
-        });
-        testSlider.data('slider')._mousedown(mouse);
-        expect(flag).toBeTruthy();
+      beforeEach(function() {
+        spy = jasmine.createSpy('spy');
       });
 
-      it("'slide' event is triggered properly and can be binded to", function() {
-        testSlider.on('slide', function() {
-          flag = true;
+      it("'slideStart' event is triggered properly and can be binded to", function(done) {
+        testSlider.on('slideStart', spy);
+        testSlider.data('slider')._mousedown(mouse);
+        setTimeout(function() {
+          expect(spy).toHaveBeenCalled();
+          done();
         });
+      });
+
+      it("'slide' event is triggered properly and can be binded to", function(done) {
+        testSlider.on('slide', spy);
         testSlider.data('slider')._mousemove(mouse);
-        expect(flag).toBeTruthy();
+        setTimeout(function() {
+          expect(spy).toHaveBeenCalled();
+          done();
+        });
       });
 
-      it("'slide' event sets the right value on the input", function() {
+      it("'slide' event sets the right value on the input", function(done) {
         testSlider.on('slide', function() {
-          flag = true;
           expect(isNaN(testSlider.val())).not.toBeTruthy();
+          done();
         });
         testSlider.data('slider')._mousemove(mouse);
-        expect(flag).toBeTruthy();
       });
 
-      it("'slide' event value and input value properties are synchronous", function() {
+      it("'slide' event value and input value properties are synchronous", function(done) {
         testSlider.on('slide', function(e) {
-          flag = true;
           expect(e.value.toString()).toEqual(this.value);
+          done();
         });
         testSlider.slider("setValue", 3, true, false);
-
-        expect(flag).toBeTruthy();
       });
 
-      it("'change' event value and input value properties are synchronous", function() {
+      it("'change' event value and input value properties are synchronous", function(done) {
         testSlider.on('change', function(e) {
-          flag = true;
           expect(e.value.newValue.toString()).toEqual(testSlider.val());
+          done();
         });
         testSlider.slider("setValue", 3, false, true);
-
-        expect(flag).toBeTruthy();
       });
 
-      it("'slideStop' event is triggered properly and can be binded to", function() {
```

Comment:
## $.trigger() is not async
I have removed the `done` argument and this test should fail because the 'slideStop' callback should be executed asynchronously. However, this test will still pass because `_trigger()` will call `$.trigger()` synchronously, and `.trigger()` will call custom events synchronously, with the exception of built-in events, which are executed sync/async depending on browser.

https://github.com/seiyria/bootstrap-slider/blob/a5fc012f717e31d292b8a837b2ba9ec9dcd8aa80/src/js/bootstrap-slider.js#L1876-L1883

See this thread: https://stackoverflow.com/questions/14017013/is-the-jquery-trigger-function-guaranteed-to-be-synchronous
```js
it("'slideStop' event is triggered properly and can be binded to", function() {
  touch.initEvent("touchstop");
  testSlider.on('slideStop', function() {  // this will be called synchronously
    expect(true).toBeTruthy();
  });
  testSlider.data('slider')._mouseup(mouse);
  window.setTimeout(function() {
    expect(spy).toHaveBeenCalled();
    // done();
  }, 0);
});
```

## Tests Should Be Async
Upon further review,

I think the issue is not that the callbacks are being executed synchronously, I think the way we had it before was fine, it's just that the **tests** themselves were not being executed asynchronously.

So I think the tests should be reorganized to be as follows:
```js
// Register callbacks
testSlider.on(/* slideStart, slide, slideStop */, function() { /* code here */});

setTimeout(function() {
  // set up code here to be executed asynchronously (next tick of the event loop)
}, 0);
```

For example,
```js
it("Should trigger change on mouseup", function(done) {
  var changes = 0;

  // Register callbacks
  testSlider.on('slideStop', function(){
    expect(changes).toBe(2);
    expect(testSlider.getValue()).toEqual([2, 5]);
    done();
  });
  
  testSlider.on('change', function(){
    changes++;
  });

  // Execute this code async (on next tick of the event loop)
  setTimeout(function() {
    testSlider.mousedown(mouseEvent('mousedown', tickOffsets[1]));
    expect(testSlider.getValue()).toEqual([1, 5]);

    testSlider.mouseup(mouseEvent('mouseup', tickOffsets[2]));
  }, 0);
});
```

If you remove the call to `done()` then the test will timeout.
```
Drag handles over each other and use keyboard to move handles over each other
     √ should drag and keydown handles properly to the right then back to the left
     √ should drag and keydown handles properly to the left then back to the right
   √ Should snap to a tick within tick bounds when using the mouse navigation
   × Should trigger change on mouseup
     Error: Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL. (1)
```